examples/07_end_to_end/seismic_waveform.ipynb

```diff
@@ -0,0 +1,1325 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "96fc1fee-0d20-45f5-8724-b6905dcbcd54",
+   "metadata": {},
+   "source": [
+    "### Real-time Adapter Definition"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "id": "4808fc0e-53a1-4677-aa28-8542ce8f31f5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import logging\n",
+    "import threading\n",
+    "from fnmatch import fnmatch\n",
+    "from obspy.clients.seedlink.client.seedlinkconnection import SeedLinkConnection\n",
+    "from obspy.clients.seedlink.client.slstate import SLState\n",
+    "from obspy.clients.seedlink.slpacket import SLPacket\n",
+    "from urllib.parse import urlparse\n",
+    "\n",
+    "from csp.impl.adaptermanager import AdapterManagerImpl\n",
+    "from csp.impl.pushadapter import PushInputAdapter\n",
+    "from csp.impl.wiring import py_push_adapter_def\n",
+    "\n",
+    "\n",
+    "class EasySeedLinkAdapter(AdapterManagerImpl):\n",
+    "    \"\"\"Real time push adapter (technically, an adapter manager) that manages subscription to multiple seedlink streams for the same connection.\n",
+    "\n",
+    "    See examples/04_writing_adapters/e7_adaptermanager_inputoutput.py for the implementation this was based on. \n",
+    "    \"\"\"\n",
+    "    \n",
+    "    def __init__(self, server_url: str):\n",
+    "        \"\"\" Constructor\n",
+    "        \n",
+    "        Args:\n",
+    "            server_url: The url and port of the seedlink server we connect to.\n",
+    "        \"\"\"\n",
+    "        self._server_url = server_url\n",
+    "        parsed_url = urlparse(server_url, scheme=\"seedlink\")\n",
+    "        self._conn = SeedLinkConnection()\n",
+    "        self._conn.set_sl_address(\"%s:%d\" % (parsed_url.hostname, parsed_url.port or 18000))\n",
+    "\n",
+    "        # We will keep track of requested input adapters here\n",
+    "        self._inputs = {}\n",
+    "\n",
+    "        # Out driving thread, all  realtime adapters will need a separate thread of execution that drives data into the engine thread\n",
+    "        self._running = False\n",
+    "        self._thread = None\n",
+    "        self._starttime = None\n",
+    "        self._endtime = None\n",
+    "\n",
+    "    def subscribe(self, network, station, selector=None, push_mode=csp.PushMode.NON_COLLAPSING):\n",
+    "        \"\"\"User facing API to subscribe to a timeseries stream from this adapter manager\"\"\"\n",
+    "        # This will return a graph-time timeseries edge representing and edge from this adapter manager for the given symbol / arguments\n",
+    "        return _easy_seedlink_adapter(self, network, station, selector, push_mode=push_mode)\n",
+    "\n",
+    "    def _create(self, engine, memo):\n",
+    "        # We'll avoid having a second class and make our AdapterManager and AdapterManagerImpl the same\n",
```

Comment:
I see youre aware of this but its bad form, I dont think we should be showcasing bad form in our examples